Make guest username, uid, home directory configurable

jandubois opened the pull request:
Also disallows the "admin" username by default (because it is a builtin user in Ubuntu), but can be overridden by setting it explicitly in lima.yaml.

```yaml
user:
  name: john
  comment: John Doe
  uid: 1000
  home: "/home/{{.User}}"
```

Another change is that `{{.User}}` and `{{.UID}}` in host templates now expand to the host values and not the guest values. Note that on Windows the host user is something like `runner\runneradmin` and host uid something like `S-1-5-21-2897396845-3571823323-1290545657-500` (on GitHub runner).

Setting `home` and `comment` does not work yet with `alpine-iso` because `lima-init` hardcodes the home directory to `/home/$USER.linux` and doesn't set the comment at all (implemented in https://github.com/lima-vm/alpine-lima/pull/134).

Fixes #2117
Fixes #2037 
Implements #2622

Replaces #2378

jandubois commented:
I recommend using "Hide whitespace" for reviewing this PR, especially for `osutil/user.go`.

afbjorklund commented:
Should there be an option to avoid "mirroring" the host values, and go with the default ones from the cloud.cfg?

afbjorklund commented on templates/default.yaml:
```diff
@@ -123,6 +123,19 @@ mountInotify: null
+  uid: null
+  # Home directory inside the VM, NOT the mounted home directory of the host.
+  # It can use the following template variables: {{.Name}}, {{.Hostname}}, {{.UID}}, {{.User}}, and {{.Param.Key}}.
+  # 🟢 Builtin default: "/home/{{.User}}.linux"
```
Should this be using {{.OS}} perhaps? (for Linux, but lowercase the value). Only makes a difference for FreeBSD

jandubois replied:
Yeah, not really part of this PR.

afbjorklund commented:
> * The code also implements support for setting `user.gid` but doesn't document it because cloud-init has no way to set it. We'll have to implement this ourselves in our ci scripts (if we want it).

In terms of features, it probably would be more useful to have "groups" (extra groups) than "gid" (not used so much)

AkihiroSuda commented on templates/default.yaml:
```diff
@@ -123,6 +123,19 @@ mountInotify: null
+user:
+  # User name. An explicitly specified username is not validated by Lima.
+  # 🟢 Builtin default: same as the host username, if it is a valid Linux username, otherwise "lima"
+  username: null
```
`user.username` stutters

afbjorklund replied:
In the code we have `user.Username` because `user.Name` is the GECOS, but the yaml can be simplified

afbjorklund replied:
Can add a separate PR for the gecos, since it seems to be used in cloud.cfg (and simple enough in ci)

afbjorklund replied:
* #2830

jandubois replied:
Ok, will go with `name`, `home`, and `uid` for now, but what should we use for `gecos`?

AkihiroSuda commented on templates/default.yaml:
```diff
@@ -123,6 +123,19 @@ mountInotify: null
+  # Home directory inside the VM, NOT the mounted home directory of the host.
+  # It can use the following template variables: {{.Name}}, {{.Hostname}}, {{.UID}}, {{.User}}, and {{.Param.Key}}.
+  # 🟢 Builtin default: "/home/{{.User}}.linux"
+  homeDir: null
```
`{{.Home}}` does not have `Dir` suffix

jandubois replied:
I've changed `homeDir` → `home`.

AkihiroSuda commented on templates/default.yaml:
```diff
@@ -123,6 +123,19 @@ mountInotify: null
 # ADVANCED CONFIGURATION
 # ===================================================================== #
 
+# User to be used inside the VM
```
Probably this should be in the `FURTHER ADVANCED CONFIGURATION`  section

jandubois replied:
done

AkihiroSuda commented on templates/default.yaml:
```diff
@@ -123,6 +123,19 @@ mountInotify: null
 # ADVANCED CONFIGURATION
 # ===================================================================== #
 
+# User to be used inside the VM
```
Can we have a test in https://github.com/lima-vm/lima/blob/master/hack/test-templates/test-misc.yaml

alexandear commented on pkg/osutil/user.go:
```diff
@@ -94,88 +92,102 @@ const (
-	sync.Once
-	u        *user.User
-	err      error
+type Cache struct {
```
`var cache struct {` was good, because no need to export `Cache`. I propose to revert. Maybe we can use `init()` functions to initialize this struct.

jandubois replied:
I'm just removing the struct altogether; it doesn't really provide anything but some namespacing that we don't really need either.

```go
var (
	once     = new(sync.Once)
	limaUser *user.User
	warnings []string
)
```

afbjorklund commented on pkg/must/must.go:
```diff
@@ -0,0 +1,8 @@
+package must
+
+func Must[T any](obj T, err error) T {
```
I feel this package to be a bit obscure, and I don't see why it should not be handled like any other error

jandubois commented:
> Should there be an option to avoid "mirroring" the host values, and go with the default ones from the cloud.cfg?

I don't know how we would set up ssh without knowing the username. If we can figure this out, then I think we should support this.

jandubois commented:
> In terms of features, it probably would be more useful to have "groups" (extra groups) than "gid" (not used so much)

I have removed all code related to `gid`, so we can figure out the correct schema in due time.

AkihiroSuda commented on pkg/limayaml/limayaml.go:
```diff
@@ -83,6 +84,13 @@ var (
 
+type User struct {
+	Username *string `yaml:"name,omitempty" json:"name,omitempty" jsonschema:"nullable"`
+	Name     *string `yaml:"gecos,omitempty" json:"gecos,omitempty" jsonschema:"nullable"`
```
The go name and the yaml name are inconsistent.

I also feel that the short name should be just called "Name" and the gecos field should be called "FullName" or maybe "Comment".

"GECOS" is a remnant from 1960s and no longer meaningful word.

jandubois replied:
I renamed it to `fullName`, but just noticed the `comment` suggestion, and will change to that one instead.

In the CI template we have `Name` for the instance name, and `User` for the user name, so `FullName` does not really fit there (so I left it at `GECOS`), but `Comment` works fine.

afbjorklund replied:
You didn't like [General Electric Comprehensive Operating Supervisor](https://en.wikipedia.org/wiki/GECOS) ? :-)

Could have gone with `Info` as well, but since nobody looks at it anyway I think `Comment` is fine...

```C
           struct passwd {
               char   *pw_name;       /* username */
               char   *pw_passwd;     /* user password */
               uid_t   pw_uid;        /* user ID */
               gid_t   pw_gid;        /* group ID */
               char   *pw_gecos;      /* user information */
               char   *pw_dir;        /* home directory */
               char   *pw_shell;      /* shell program */
           };
```

Apparently I had some bugs with gecos in "tinycloudinit", and currently it uses `Linux User,,,`

https://github.com/mirror/busybox/blob/1_36_1/loginutils/adduser.c#L199

AkihiroSuda reviewed (changes requested):
The go name and the yaml name are inconsistent.

AkihiroSuda commented on templates/default.yaml:
```diff
@@ -279,6 +279,19 @@ containerd:
+user:
+  # User name. An explicitly specified username is not validated by Lima.
+  # 🟢 Builtin default: same as the host username, if it is a valid Linux username, otherwise "lima"
+  name: null
```
Do you want to add `fullName` too?

AkihiroSuda commented on pkg/limayaml/limayaml.go:
```diff
@@ -83,6 +84,13 @@ var (
 )
 
+type User struct {
+	Username *string `yaml:"name,omitempty" json:"name,omitempty" jsonschema:"nullable"`
```
The go name and the yaml name should be synced

jandubois commented:
I'm wondering if `{{.Comment}}` should be supported in host/guest templates. It would be easy to add, but I can't think of an actual use-case.

jandubois commented:
I believe I've addressed all feedback now, PTAL

There are 2 open issues that should probably be done "later":

* Optionally don't add an additional user at all (but how do we set up ssh keys?)
* Add `{{.Comment}}` to host and guest templates (but why?)

AkihiroSuda commented on pkg/limayaml/limayaml.go:
```diff
@@ -83,6 +84,13 @@ var (
+	Name    *string `yaml:"name,omitempty" json:"name,omitempty" jsonschema:"nullable"`
+	Comment *string `yaml:"comment,omitempty" json:"comment,omitempty" jsonschema:"nullable"`
+	Home    *string `yaml:"home,omitempty" json:"home,omitempty" jsonschema:"nullable"`
+	UID     *string `yaml:"uid,omitempty" json:"uid,omitempty" jsonschema:"nullable"`
```
Why not `*int`?

jandubois replied:
No particular reason, I can change it.

jandubois replied:
@AkihiroSuda Can we leave this for a later refactoring PR? I would also want to combine this type with https://github.com/lima-vm/lima/blob/9fe06aa4cc7ac7d1d6b72787301753cc324ae4d8/pkg/osutil/user.go#L17-L24

But I feel like the current PR is already big enough.

AkihiroSuda commented on pkg/cidata/template_test.go:
```diff
@@ -30,11 +30,11 @@ func TestConfig(t *testing.T) {
-		Home:  "/home/foo.linux",
+		Name:    "default",
+		User:    "foo",
+		UID:     501,
```
Does the yaml library support unmarshalling this number literal to *string?

LGTM then, feel free to merge it

jandubois replied:
Somewhat ironically the `UID` field in the cidata template is an `int`, and then gets turned into a string again in `user-data` with `uid: "{{.UID}}"`. 😄 

Anyways, I've change `uid` to an `int` in `lima.yaml` because I don't like changing the schema again later.

So please approve again once everything is green again.

jandubois replied:
Ok, changed it one more time to `uint32` to avoid CodeQL errors...

jandubois replied:
> LGTM then, feel free to merge it

You have requested changes; please formally approve. I don't want to merge without the explicit approval because it would hurt our OSSF score...

AkihiroSuda reviewed (approved):
Thanks

afbjorklund commented:
> I'm wondering if `{{.Comment}}` should be supported in host/guest templates. It would be easy to add, but I can't think of an actual use-case.

I don't actually know the use case (even `finger` has been removed, for "pinky") - it was just that all the cloud-config examples had gecos...
